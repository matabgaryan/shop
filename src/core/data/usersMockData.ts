import {Role} from "../enums";
import {Pojo} from "../types";

export const usersMockData: Pojo = {
    user: {
        id: '1',
        username: 'user',
        email: 'user@exampl.com',
        firstName: 'John',
        lastName: 'Doe',
        roles: [Role.USER],
    },
    admin: {
        id: '2',
        username: 'admin',
        email: 'admin@exampl.com',
        firstName: 'admin',
        lastName: 'Doe',
        roles: [Role.ADMIN, Role.USER],
    },
}
