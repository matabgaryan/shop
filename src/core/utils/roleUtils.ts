import {Role} from "../enums";

export const checkUserRole = (routeRoles: Role[], authenticatedUserRoles: Role[]) =>  {
    return routeRoles.some(item => authenticatedUserRoles.includes(item))
}
