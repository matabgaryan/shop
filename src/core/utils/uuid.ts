import { v4 as uuidv4 } from "uuid";

export const generatedUuid: () => string = () => {
    return uuidv4();
}
