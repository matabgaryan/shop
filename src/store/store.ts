import {combineReducers, configureStore, PreloadedState} from "@reduxjs/toolkit";
import shopCart from "./slices/shopCart";
import productsReducer from "./slices/products";
import authReducer from './slices/auth';
import modalReducer from './slices/modals';
import {RootState} from "../core/types";

export const store = configureStore({
    reducer:{
        shopCart,
        products: productsReducer,
        auth: authReducer,
        modals: modalReducer,

    },
    devTools: true,
})
const rootReducer = combineReducers({
    shopCart,
    products: productsReducer,
    auth: authReducer,
    modals: modalReducer,
});


export function setupStore(preloadedState?: PreloadedState<RootState>) {
    return configureStore({
        reducer: rootReducer,
        preloadedState
    })
}

