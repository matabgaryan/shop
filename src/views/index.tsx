export * from './shared/NotFound';
export * from './shared/AccessDenied';
export * from './shared/Home';
export * from './shared/Products';
export * from './admin'
