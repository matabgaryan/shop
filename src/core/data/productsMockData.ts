import {generatedUuid} from "../utils/uuid";
import {IProduct} from "../../store/slices/interface";

export const productsData: IProduct[] = [
    {
        id: generatedUuid(),
        name: 'Product 1',
        price: '30',
        weight: 100,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 2',
        price: '30',
        weight: 100,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 3',
        price: '30',
        weight: 20,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 4',
        price: '30',
        weight: 30,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 5',
        price: '30',
        weight: 40,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 6',
        price: '30',
        weight: 200,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 7',
        price: '30',
        weight: 50,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 8',
        price: '30',
        weight: 100,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 9',
        price: '30',
        weight: 65,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
    {
        id: generatedUuid(),
        name: 'Product 10',
        price: '30',
        weight: 70,
        startDate: 'May 5 , 2022',
        endDate: 'Jun 5 2022',
        count: 0
    },
]
