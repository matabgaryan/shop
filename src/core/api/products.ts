import { productsData } from '../data/productsMockData'
import {IProduct} from "../../store/slices/interface";

export const getProductsList: () => Promise<IProduct[]> = () => {
    return  new Promise((resolve) => {
        setTimeout(() => resolve(productsData), 250);
    });
}
