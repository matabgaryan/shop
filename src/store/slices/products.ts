import {ActionReducerMapBuilder, createAsyncThunk, createSlice, SliceCaseReducers} from "@reduxjs/toolkit";
import {STATUS} from "../../core/enums";
import {getProductsList} from "../../core/api/products";
import {IProduct, IProductsState} from "./interface";
import {NoInfer} from "@reduxjs/toolkit/dist/tsHelpers";
import {RootState} from "../../core/types";

const initialState: IProductsState = {
    status: "",
    products: [],
};

const productSlice = createSlice<IProductsState, SliceCaseReducers<IProductsState>>({
    name: "products",
    initialState,
    reducers: {
        addProducts: (state, action) => {
            state.products.push(...action.payload);
        }
    },
    extraReducers: (builder: ActionReducerMapBuilder<NoInfer<IProductsState>>) => {
        builder
            .addCase(getProducts.pending, (state) => {
                state.status = STATUS.PENDING;
            })
            .addCase(getProducts.fulfilled, (state, action) => {
                state.products = action.payload;
                state.status = STATUS.SUCCEEDED;
            })
            .addCase(getProducts.rejected, (state) => {
                state.status = STATUS.REJECTED;
            });
    },
});

export const getProducts = createAsyncThunk<IProduct[]>("fetch/products", async () => {
    return await getProductsList();
});
export const { addProducts } = productSlice.actions;

export const filteredProducts = (state: RootState) => state.products.products

export default productSlice.reducer;
