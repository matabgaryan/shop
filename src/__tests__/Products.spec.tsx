import {
    render,
    screen,
    fireEvent,
    RenderResult, waitFor,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import {Products} from "../views";
import {setupStore, store} from '../store/store';
import {renderWithProviders} from "../core/utils/testUtils";
import {productsData} from "../core/data/productsMockData";
import { addProducts } from "../store/slices/products";

const renderProducts = (): RenderResult =>
    render(
        <Provider store={store}>
               <MemoryRouter>
                   <Products  />
               </MemoryRouter>
        </Provider>
    );
describe('Products', () => {
    test('Renders Init Products', () => {
        const { asFragment } = renderProducts();
        expect(asFragment()).toMatchSnapshot();
    });

    test('Should render products list', async () => {
        const store = setupStore();
        store.dispatch(addProducts(productsData));
        renderWithProviders( <Products  />, {store})
        const initialLength = store.getState().products.products.length;
        expect(initialLength).toEqual(10);
        const allProductsBtn = screen.getByTestId('products-show-all-btn');
        const productsFilterBtn = screen.getByTestId('products-filter-btn');
        expect(allProductsBtn).toBeInTheDocument();
        expect(productsFilterBtn).toBeInTheDocument();
        expect(allProductsBtn.classList.contains('bg-white')).toBe(false);
        expect(productsFilterBtn.classList.contains('bg-appBlack')).toBe(false)
        const productList = screen.getByTestId('products-list');
        const productItem = screen.getAllByTestId('product-item');
        expect(productList).toBeInTheDocument();
        expect(productItem).toHaveLength(10);
        fireEvent.click(allProductsBtn);
        await waitFor(() => {
            expect(allProductsBtn.classList.contains('bg-appBlack')).toBe(true);
        })
    });
})

