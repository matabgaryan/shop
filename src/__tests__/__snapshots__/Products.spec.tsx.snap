// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products Renders Init Products 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="w-full flex flex-col items-center"
    >
      <div
        class="w-full"
      >
        <p
          class="text-appBlack font-GoogleSansBold font-[700] text-[22px]"
        >
          Shop by Collection
        </p>
        <p
          class="max-w-2xl font-GoogleSansRegular font-[400] text-[18px] text-appBlack"
        >
          Each season, we collaborate with world class designers to create a collection inspired by natural world.
        </p>
      </div>
      <div
        class="w-full flex items-center justify-end"
      >
        <button
          class="bg-white text-appBlack py-2.5 px-5 mr-2 mb-2 text-sm font-medium  focus:outline-none rounded-lg border border-gray-200 "
          data-testid="products-filter-btn"
          type="button"
        >
          More Probable Products
        </button>
        <button
          class="bg-appBlack text-white py-2.5 px-5 mr-2 mb-2 text-sm font-medium focus:outline-none  rounded-lg border border-gray-200  "
          data-testid="products-show-all-btn"
          type="button"
        >
          All Products
        </button>
      </div>
      <div
        class="flex flex-col justify-center items-center min-h-[50%]			"
      >
        <div
          class="flex space-x-1 mt-3  rounded-full justify-center items-center"
        >
          <div
            class="bg-appBlack  w-2 h-2 rounded-full animate-pulse blue-circle"
          />
          <div
            class="bg-appBlack  w-2 h-2 rounded-full animate-pulse green-circle"
          />
          <div
            class="bg-appBlack  w-2 h-2 rounded-full animate-pulse red-circle"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
